Xray comparisons raised from the dead

sbelak opened the pull request:
TODO
- [x] API endpoints [done in the context of xrays. Will need some thinking to open it for other dashboards]
- [x] tests
- [x] UI entry points
- [x] Add related sidebar

sbelak committed:
Reenable comparisons

sbelak committed:
Merge remote-tracking branch 'origin/master' into xray-comparisons-raise-from-the-dead

sbelak committed:
Color by segment, add segment names to legends

sbelak committed:
Merge branch 'master' into xray-comparisons-raise-from-the-dead

sbelak committed:
Merge remote-tracking branch 'origin/xray-api-add-permission-checks' into xray-comparisons-raise-from-the-dead

sbelak committed:
Add provisional API endpoint and UI hook via related

sbelak commented:
Added provisional entry point via related:
<img width="448" alt="screenshot 2018-06-12 20 01 11" src="https://user-images.githubusercontent.com/178038/41308172-7cfdfdae-6e7b-11e8-89ae-5c8dfddb109a.png">
<img width="1419" alt="screenshot 2018-06-12 20 01 02" src="https://user-images.githubusercontent.com/178038/41308179-8125bbd8-6e7b-11e8-98c7-75c4758ee9e3.png">

sbelak committed:
Reorder defns

sbelak committed:
Remove unneeded requires

sbelak committed:
Ensure all segment components are present in the comparison

sbelak committed:
Merge branch 'xray-fix-aggregate-card-fks' into xray-comparisons-raise-from-the-dead

sbelak commented:
We now always show all the fields included in the segments. This is done by composing given dashboard with a dashboards for a special comparison heuristic. This allows us to customize what's shown for every field type x table type combination.

mazameli commented:
So I just tried this with two segments:
- "Large orders," where subtotal > 100
- "Tiny orders," where subtotal < or = to 20

When I x-ray `large orders` and then compare that to `tiny orders`, I'm getting incorrect results. `Tiny orders` should have 34 rows. What appears to be happening is that the filter in the definition for `Large orders` is incorrectly being applied to the cards about `tiny orders`, so that it's trying to show me orders where subtotal is > 100 AND < or = to 20 (which of course don't exist).

<img width="1328" alt="screen shot 2018-06-14 at 12 08 37 pm" src="https://user-images.githubusercontent.com/2223916/41432850-09b37be2-6fcc-11e8-8e2a-535c411eb23b.png">

Here's what I see when I clicked through on the top-right, total transactions card:

<img width="976" alt="screen shot 2018-06-14 at 12 08 47 pm" src="https://user-images.githubusercontent.com/2223916/41432856-0d7c05c8-6fcc-11e8-8c53-1bcb769d3ab1.png">

sbelak committed:
Don't inherit `query-filter`

sbelak commented:
Good catch. Fixed

sbelak committed:
Change api endpoint argument name to avoid broken magic

sbelak committed:
Manually cast ids to int

sbelak committed:
Don't try to recast ints to int

sbelak committed:
Don't split Y axis for overlay cards

sbelak committed:
Add segment descriptions where possible. Move segment summary up top

sbelak committed:
Add backend tests, factor out common utilites for testing xrays

sbelak committed:
Merge branch 'master' into xray-comparisons-raise-from-the-dead

sbelak committed:
Add frontend tests

sbelak committed:
Add "compare with entire dataset" related for segments.

sbelak committed:
Account for description row height

sbelak committed:
Add related

sbelak commented:
Added: 
* descriptions to title row (where applicable)
* sidebar with related

Moved comparison of fields used in definition of segment (or `:filter`) to the top.

<img width="1403" alt="screenshot 2018-06-20 00 34 09" src="https://user-images.githubusercontent.com/178038/41627811-38ac199a-7422-11e8-97ed-0b35cb97cb60.png">

sbelak committed:
Don't use u/get-id where we might be dealing with adhoc queries

sbelak committed:
fix indentation

sbelak committed:
Improve copy. Make cell titles more sensible

sbelak committed:
Split cases where we have query-filter and cell-query

sbelak committed:
Introduce [[this.foo]] syntax and make GenericTable and segments nice

sbelak committed:
Alow acces to both root and what's binded to this via dot notation

sbelak committed:
Don't stuff filter clauses into `cell-query`.

mazameli committed:
small tweaks and fixes

sbelak committed:
fix typo short-title -> short-name

sbelak committed:
Change cell xray to be based off card xray

sbelak committed:
downsize titles when merging dashboards

sbelak committed:
Fix enumeration of field names in titles

sbelak committed:
Correctly handle cell xrays of questions with multiple breakout dims

sbelak committed:
Fix regression in candidates

sbelak committed:
Fix broken url generation

sbelak committed:
Test for more working correctly

sbelak committed:
Don't show cards and filters where breakdown dim is a cell dim

sbelak committed:
don't explode if cell-query is nil

sbelak committed:
Correct title for questions with no breakout dimension

sbelak committed:
tests: set up collection for cards so that permissions work

salsakran committed:
entry action

tlrobinson commented on frontend/src/metabase/qb/components/actions/CompareWithTable.jsx:
```diff
@@ -0,0 +1,38 @@
+      url: () =>
+        question.card().id
+          ? `/api/automagic-dashboards/table/${tableId}/compare/question/${
+              question.card().id
```
I think this can just be `question.id()`

tlrobinson commented on frontend/src/metabase/qb/components/actions/CompareWithTable.jsx:
```diff
@@ -0,0 +1,38 @@
+              question.card().id
+            }`
+          : `/api/automagic-dashboards/table/${tableId}/compare/adhoc/${utf8_to_b64url(
+              JSON.stringify(question.card().dataset_query),
```
We should add a method to `Question` to get the serialized query, but this is ok for now.

mazameli committed:
remove use of 'these'

mazameli committed:
merge changes

sbelak committed:
Appease the defendpoint gods

sbelak committed:
explicitly test for show=1

salsakran committed:
update url

salsakran committed:
drill entry points

sbelak committed:
Don't throw away :query-filter for adhocs; expand segment macros

sbelak commented:
There is slight weirdness with titles if you compare to an adhoc that's effectively a filter. eg:
<img width="1081" alt="screenshot 2018-06-29 13 06 18" src="https://user-images.githubusercontent.com/178038/42089464-8755d414-7b9d-11e8-877b-d69169baaae3.png">

Not entirely sure what the best course of action here would be.

sbelak committed:
Add docstring

sbelak committed:
Don't skip small categories in GenericTable

sbelak committed:
correctly flow query-filter to aggregate card constituents

sbelak committed:
more robust save dashboard test

sbelak committed:
Fix nested cards and improve filter merging

sbelak committed:
Fix a bunch of bugs

sbelak committed:
remove unneeded require

sbelak committed:
Flatten filters where possible

sbelak committed:
Discard cell-query for comparison side

sbelak commented:
Cell comparisons now do the sensible thing so you can eg. compare a cell to all of it's kin by:
/adhoc/.../cell/.../compare/table/X

sbelak commented:
Before this gets merged we have to figure out what to do with the sidebar. As is there is way too many entries in it.

mazameli commented:
I'll take a look at this branch and the sidebar.

mazameli commented:
Hmm, @sbelak, when I try clicking on the `compare these to the rest` cell x-ray action I end up on the 404 page and see `API endpoint does not exist` in the server logs.

sbelak committed:
upgrade all calls to the new api

sbelak commented:
Wrong number of args (3)  -- fixed

Can you describe the steps that get you 404 more please

mazameli commented:
Sure, but it's nothing fancy. Just e.g. clicking on this link in this example:

![screen shot 2018-07-03 at 3 41 06 pm](https://user-images.githubusercontent.com/2223916/42247866-9a22f774-7ed7-11e8-8453-6d4c2b693112.png)

sbelak committed:
Fix drill through links

sbelak commented:
Ugh, I was just being thick and didn't have the latest FE version running. Fixed.

mazameli commented:
Comparing these Doohickey orders to the rest doesn't seem to do what I expect it to.

![screen shot 2018-07-03 at 5 00 05 pm](https://user-images.githubusercontent.com/2223916/42249900-9bf0f424-7ee2-11e8-9aae-9e4f42986ac0.png)

It looks like it's just showing me all orders vs. all orders:

![screen shot 2018-07-03 at 4 59 53 pm](https://user-images.githubusercontent.com/2223916/42249944-d94aada6-7ee2-11e8-90e9-b1298ad765cc.png)

sbelak committed:
Fix cell comparisons. This time for real :(

sbelak committed:
Flatten filter clause

sbelak commented:
Ugh, fixed, tnx

sbelak committed:
Don't show superfluous datetime components

sbelak committed:
update tests

mazameli committed:
fix tiny typo

mazameli committed:
reorder minute and hour titles, remove comma from year titles

sbelak committed:
Humanize periodical datetime components nicer

mazameli commented:
Right now comparison x-rays are a big ol' wall of charts, and I think we should try breaking up the sections with some centered subheadings, something like this (ignore the copy itself; I think I have a better idea for copy at the bottom of this comment):

<img width="1398" alt="screen shot 2018-07-05 at 3 31 52 pm" src="https://user-images.githubusercontent.com/2223916/42351363-7560c5fe-8069-11e8-84a5-b377b11a0ee7.png">


I mocked that up just within Metabase itself with these text card settings:

<img width="1380" alt="screen shot 2018-07-05 at 3 32 12 pm" src="https://user-images.githubusercontent.com/2223916/42351372-870c88d8-8069-11e8-8bdf-d34c2f06125f.png">
<img width="1181" alt="screen shot 2018-07-05 at 3 32 24 pm" src="https://user-images.githubusercontent.com/2223916/42351374-89c3c4c4-8069-11e8-8d4e-565a4ca75c93.png">


**As for the sections themselves, it seems like these are the broad sections of cards:**
- [Time] "How they compare over time"
- [Categories] "How they compare across different categories"
- [Location] "How their locations compare"

sbelak commented:
Currently we throw away the structure of the source dashboard. Can we reuse that?

sbelak committed:
Fix related

sbelak committed:
fix regression

sbelak committed:
Properly handle timezone

sbelak committed:
Add type hint

sbelak committed:
fix wrong arg order. ADD TESTS for datetime humanization

sbelak committed:
Fix type hint

mazameli commented:
> Currently we throw away the structure of the source dashboard. Can we reuse that?

I don't know. Want to try it and then we can see how it looks if it's easy enough to do?

mazameli committed:
fix dangling  in day-of-week xrays

sbelak committed:
Add day-of-year and test for all the available units

sbelak committed:
Retain structure (provisional)

sbelak commented:
MVP structure. The texts should probably be something different, but at least we have something to start thinking around

mazameli commented:
Cool, so I see the new headings that have appeared. How much can we alter the text in these headings when they're in the comparison context? I.e. I want these headings to stay what they are for single-segment etc. x-rays, but I want them to be totally different text when they're part of a comparison x-ray (and ideally they'd be centered to make it clearer that the heading is about both things being compared).

E.g in a comparison x-ray between a segment and the whole table, I would want a heading that currently says "A look at your Salary field" to instead say something like "How they compare across the Salary field."

sbelak commented:
I'll add a new field to the group specification in the heuristic and we'll have full control over it.

sbelak committed:
Update test

mazameli commented:
Awesome. One unrelated thing I noticed that changed is that these titles are misaligned now (they should both begin at the same y position):

![screen shot 2018-07-09 at 1 33 14 pm](https://user-images.githubusercontent.com/2223916/42476449-d936353e-8382-11e8-92ca-66c0642dba2e.png)

mazameli committed:
address sameer's feedback

sbelak committed:
Merge branch 'master' into xray-copy-tweaks

sbelak committed:
Fix bad merge

sbelak committed:
Add comparison_title

sbelak committed:
Merge branch 'xray-copy-tweaks' into xray-comparisons-raise-from-the-dead

sbelak commented:
I've added a new field comparison_title to groups (see GenericTable.yaml for example) which is used when showing comparisons. 

Also provisionally merged copy tweaks.

sbelak commented:
What have we decided abut the sidebar?

mazameli commented:
> What have we decided abut the sidebar?

My assumption is that we'll implement this: https://github.com/metabase/metabase/issues/7983

mazameli commented:
**Quick thoughts on titles of "ad hoc" segments in drill-through comparisons:**

If I'm looking at orders by day and I click on this bar to compare it to the rest of the orders:

![screen shot 2018-07-12 at 9 59 27 am](https://user-images.githubusercontent.com/2223916/42648396-a2223d28-85bb-11e8-8aef-2731f8a9297c.png)

I'd expect the title of the ad hoc segment on the left to read `Number of Orders on June 30, 2018`. (I.e. the aggregation plus the cell's filter value(s).) The one on the right would be better as `All Orders`.

![screen shot 2018-07-12 at 10 00 04 am](https://user-images.githubusercontent.com/2223916/42648407-accf23da-85bb-11e8-9efe-8490235e5f34.png)

Similarly, I think these card titles would read better if they were `Number of Orders on June 30, 2018`, and `All Orders`. Or for another example, if the original aggregation was average of quantity, then they'd be `Average of Quantity on June 30, 2018` and `Average of Quantity, all Orders`.

![screen shot 2018-07-12 at 10 00 19 am](https://user-images.githubusercontent.com/2223916/42648578-3829b648-85bc-11e8-886a-5f6a3a64f3db.png)

mazameli commented:
I saw you added `comparison_title` to GenericTable.yaml, which is great. Are we also going to have control over titles in cell/drillthrough comparisons?

mazameli committed:
better distribution comparison title

mazameli commented:
**Summary of all the remaining things I see:**

- [ ] Titles of compared entities should always top-align, even if one doesn't have a description

![screen shot 2018-07-12 at 10 34 08 am](https://user-images.githubusercontent.com/2223916/42649746-594e31fc-85bf-11e8-9bd1-41c38921eeb4.png)

- [ ] Ad hoc segment titles in drill-through comparisons should make sense ([see comment above](https://github.com/metabase/metabase/pull/7792#issuecomment-404586141)).
- [ ] In cell comparisons, `How this metric is distributed across different numbers/categories` should read `How they're distributed across different numbers/categories`.

- [ ] The `A look at your x table` subheading in comparisons is unnecessary.

![screen shot 2018-07-12 at 10 36 50 am](https://user-images.githubusercontent.com/2223916/42649828-90ce3690-85bf-11e8-9baa-d36b3ae3b808.png)

- [ ] Cell x-rays have an unnecessary subheading here that's taking up a lot of space:

![unnecessary](https://user-images.githubusercontent.com/2223916/42650273-d02822fa-85c0-11e8-8217-52fd129936f6.png)

sbelak commented:
Yes, comparison_title is available in all heuristics, with the same [current] limitation as groups -- for cells they will be inherited from Metrics and Fields groups.

sbelak committed:
Top align titles

sbelak committed:
Better cell titles; skip redundant titles

sbelak committed:
Add comparison_title to heursitics

sbelak commented:
Done, done, done & done

sbelak committed:
Make flow happy

sbelak committed:
Generate correct title for cells of row questions

mazameli commented:
Only ugly thing I just came across is the case when there are a number of different time fields in a table/segment, it causes a bunch of scalar cards to show up without any indication of which time field is being used. Maybe in this situation with multiple time fields we do something like `Salaries in the last 30 days (by Birth Date)`.

<img width="1105" alt="screen shot 2018-07-23 at 4 26 03 pm" src="https://user-images.githubusercontent.com/2223916/43101214-cf4824b0-8e95-11e8-8c61-7ad85b63088e.png">

----

Also, IMO we need to drop these redundant `Blahs in the` before the name of the segments in titles and in the heading at the very top of the comparison. This is based on the assumption that names+descriptions of segments are going to be enough to understand what they represent without needing the name of the table they're based off of. It just seems pretty badly redundant to me like this:

<img width="1043" alt="screen shot 2018-07-23 at 4 34 40 pm" src="https://user-images.githubusercontent.com/2223916/43101668-e7dc6d00-8e96-11e8-80fe-1a9f34c5d540.png">

----

Another small thing that did throw me was the use of split axes when the delta between two timeseries was large. Feels like it makes comparing things tough.

<img width="1104" alt="screen shot 2018-07-23 at 4 17 18 pm" src="https://user-images.githubusercontent.com/2223916/43100918-d3df774a-8e94-11e8-9357-b3aa7c77022b.png">

sbelak committed:
Shorter names for segments, don't split Y axis in overlays
